pkg/tfgen/internal/unrec/recursion_detector.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unrec
+
+import (
+	"sort"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+// Detects unrolled recursion.
+//
+// TF does not support recursive type definitions but Pulumi does. The idea here is to take a projection of TF recursion
+// unrolling trick and fold it back.
+//
+// Suppose ancestor(T1, T2) if T2 describes a sub-property of T1.
+//
+// Suppose T1 <= T2 if T1 has the subset of properties of T2 with types that are themselves Tx <= Ty.
+//
+// The algo looks for this pattern to detect recursion:
+//
+//	ancestor(T1, T2), ancestor(T2, T3),  T3 <= T2 <= T1
+//
+// This needs to use an approximate and not strict equality because the leaf node of a recursively unrolled type will
+// drop recursive properties and therefore not strictly match the ancestor.
+type recursionDetector struct {
+	schema                 *pschema.PackageSpec
+	detectedRecursiveTypes map[tokens.Type]struct{}
+	cmp                    *comparer
+}
+
+func newRecursionDetector(schema *pschema.PackageSpec) *recursionDetector {
+	return &recursionDetector{
+		schema:                 schema,
+		detectedRecursiveTypes: map[tokens.Type]struct{}{},
+		cmp:                    &comparer{schema: schema},
+	}
+}
+
+func (rd *recursionDetector) Detect(roots []tokens.Type) []tokens.Type {
```

Comment:
Is `roots` a resource token or a type token?

Reply:
Always a type token.